pkg/sql/schemachanger/schemachanger_test: TestExecuteWithDMLInjection_alter_table_add_unique_without_index failed

pkg/sql/schemachanger/schemachanger_test.TestExecuteWithDMLInjection_alter_table_add_unique_without_index [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/11401393?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/11401393?buildTab=artifacts#/) on master @ [31fe139c74640a383599c53719a39f0190a7065d](https://github.com/cockroachdb/cockroach/commits/31fe139c74640a383599c53719a39f0190a7065d):


```
        	            				github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest/cumulative.go:242
        	            				github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest/cumulative.go:153
        	Error:      	Received unexpected error:
        	            	query 'SELECT status, job_type, description FROM [SHOW JOBS] WHERE job_type IN ('SCHEMA CHANGE', 'TYPEDESC SCHEMA CHANGE', 'NEW SCHEMA CHANGE') AND status NOT IN ('succeeded', 'failed')': expected:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:315
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:75
        	            	  | github.com/cockroachdb/cockroach/pkg/util/retry.ForDuration
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithinError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:81
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).succeedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:117
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.(*SQLRunner).CheckQueryResultsRetry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/sql_runner.go:312
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest.waitForSchemaChangesToFinish
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest/framework.go:932
        	            	  | runtime.Goexit
        	            	  | 	GOROOT/src/runtime/panic.go:522
        	            	  | testing.(*common).FailNow
        	            	  | 	GOROOT/src/testing/testing.go:871
        	            	  | testing.(*common).Fatalf
        	            	  | 	GOROOT/src/testing/testing.go:955
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsWithin
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:64
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon
        	            	  | 	github.com/cockroachdb/cockroach/pkg/testutils/soon.go:42
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest.executeSchemaChangeTxn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest/framework.go:885
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest.ExecuteWithDMLInjection.func2.2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest/cumulative.go:224
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest.SingleNodeTestClusterFactory.Run
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest/test_server_factory.go:105
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest.ExecuteWithDMLInjection.func2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest/cumulative.go:242
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest.ExecuteWithDMLInjection.func1.2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/sql/schemachanger/sctest/cumulative.go:153
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) query 'SELECT status, job_type, description FROM [SHOW JOBS] WHERE job_type IN ('SCHEMA CHANGE', 'TYPEDESC SCHEMA CHANGE', 'NEW SCHEMA CHANGE') AND status NOT IN ('succeeded', 'failed')': expected:
        	            	  |
        	            	  | got:
        	            	  | running, NEW SCHEMA CHANGE, ALTER TABLE defaultdb.public.t ADD CONSTRAINT unique_j UNIQUE WITHOUT INDEX (j)
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
    --- FAIL: TestExecuteWithDMLInjection_alter_table_add_unique_without_index/injection_stage_{1_2_3_false}#01 (92.65s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExecuteWithDMLInjection_alter_table_add_unique_without_index.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30774

--- a/pkg/sql/schemachanger/sctest/cumulative.go
+++ b/pkg/sql/schemachanger/sctest/cumulative.go
@@ -121,7 +121,6 @@ func ExecuteWithDMLInjection(t *testing.T, relPath string, factory TestServerFac
 	var testDMLInjectionCase func(
 		t *testing.T, ts CumulativeTestSpec, key stageKey, injectPreCommit bool,
 	)
-	var injectionFunc execInjectionCallback
 	testFunc := func(t *testing.T, ts CumulativeTestSpec) {
 		// Count number of stages in PostCommit and PostCommitNonRevertible phase
 		// for running `stmts` after properly running `setup`.
@@ -136,7 +135,6 @@ func ExecuteWithDMLInjection(t *testing.T, relPath string, factory TestServerFac
 				}
 			}
 		})
-		injectionFunc = ts.stageExecMap.GetInjectionCallback(t, ts.Rewrite)
 		injectionRanges := ts.stageExecMap.GetInjectionRuns(postCommit, nonRevertible)
 		defer ts.stageExecMap.AssertMapIsUsed(t)
 		injectPreCommits := []bool{false}
@@ -167,6 +165,8 @@ func ExecuteWithDMLInjection(t *testing.T, relPath string, factory TestServerFac
 		var clusterCreated atomic.Bool
 		var tdb *sqlutils.SQLRunner
 		ctx := context.Background()
+		injectionFunc := ts.stageExecMap.GetInjectionCallback(t, ts.Rewrite)
+
 		knobs := &scexec.TestingKnobs{
 			BeforeStage: func(p scplan.Plan, stageIdx int) error {
 				if !clusterCreated.Load() {
@@ -175,6 +175,19 @@ func ExecuteWithDMLInjection(t *testing.T, relPath string, factory TestServerFac
 					// defaultdb`) and we don't want those to hijack this knob.
 					return nil
 				}
+
+				// if t.Fail/Fatal has been called, try to end the job as quickly as
+				// possible and avoid running any more test code.
+				if t.Failed() && p.InRollback {
+					t.Log("short-circuiting BeforeStage hook due to test failure")
+					return nil
+				}
+
+				if t.Failed() {
+					t.Log("forcing job failure from BeforeStage due to test failure")
+					return errors.New("t.Failed() is true")
+				}
+
 				s := p.Stages[stageIdx]
 				if (injection.phase == p.Stages[stageIdx].Phase &&
 					p.Stages[stageIdx].Ordinal >= injection.minOrdinal &&
@@ -207,11 +220,11 @@ func ExecuteWithDMLInjection(t *testing.T, relPath string, factory TestServerFac
 					} else {
 						t.Logf("Retrying stage: %v", key)
 					}
-
 				}
 				return nil
 			},
 		}
+
 		runfn := func(s serverutils.TestServerInterface, db *gosql.DB) {
 			clusterCreated.Store(true)
 			tdb = sqlutils.MakeSQLRunner(db)

--- a/pkg/sql/schemachanger/sctest/framework.go
+++ b/pkg/sql/schemachanger/sctest/framework.go
@@ -465,7 +465,7 @@ func (e *stageExecStmt) Exec(
 			if err != nil {
 				if idx != len(e.stmts)-1 {
 					// We require that only the last statement in a stage-exec block can cause an error.
-					t.Fatalf("unexpected error encountered; only the last statement can cause an error")
+					t.Fatalf("unexpected error (only the last statement may expect an error): %v", err)
 				} else {
 					// Fail the test unless the error is expected (from e.expectedOutput), or
 					// "rewrite" is set, in which case we record the error and proceed.
@@ -487,8 +487,14 @@ func (e *stageExecStmt) Exec(
 			results := runner.QueryStr(t, boundSQL)
 			if !reflect.DeepEqual(results, expectedQueryResult) {
 				if !rewrite {
-					t.Fatalf("query '%s': expected:\n%v\ngot:\n%v\n",
-						stmt, sqlutils.MatrixToStr(expectedQueryResult), sqlutils.MatrixToStr(results))
+					t.Fatalf(
+						"query '%s' ($stageKey=%d,$successfulStageCount=%d): expected:\n%v\ngot:\n%v\n",
+						stmt,
+						stageVariables.stage.AsInt()*1000,
+						stageVariables.successfulStageCount,
+						sqlutils.MatrixToStr(expectedQueryResult),
+						sqlutils.MatrixToStr(results),
+					)
 				}
 				e.expectedOutput = sqlutils.MatrixToStr(results)
 			}
@@ -888,7 +894,7 @@ func executeSchemaChangeTxn(
 			err = e
 			return nil
 		default:
-			return errors.New("waiting for statements to execute")
+			return errors.Newf("waiting for statements to execute: %v", spec.Stmts)
 		}
 	})
 	return err

